File: plot_trajectory.py
import os
import glob
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime

def list_gps_files():
    """List all GPS data files in the gps_data directory"""
    data_dir = 'gps_data'
    if not os.path.exists(data_dir):
        print(f"Error: Directory '{data_dir}' not found.")
        return []
    
    files = glob.glob(os.path.join(data_dir, "gps data *.txt"))
    files.sort(key=os.path.getmtime)  # Sort by modification time
    return files

def get_latest_file():
    """Get the most recently created GPS data file"""
    files = list_gps_files()
    if not files:
        return None
    return files[-1]  # Return the most recent file

def select_file():
    """Allow user to select a file to plot"""
    files = list_gps_files()
    
    if not files:
        print("No GPS data files found.")
        return None
    
    print("\nAvailable GPS data files:")
    for i, file in enumerate(files):
        filename = os.path.basename(file)
        print(f"{i+1}. {filename}")
    
    print(f"\n{len(files)+1}. Latest file ({os.path.basename(files[-1])})")
    
    try:
        choice = int(input(f"\nSelect file to plot (1-{len(files)+1}), or press Enter for latest: ") or len(files)+1)
        if choice == len(files)+1:
            return files[-1]
        elif 1 <= choice <= len(files):
            return files[choice-1]
        else:
            print("Invalid selection.")
            return None
    except ValueError:
        print("Invalid input.")
        return None

def plot_trajectory(file_path):
    """Plot GPS trajectory from the specified file"""
    try:
        # Read the data file
        data = pd.read_csv(file_path)
        
        # Extract file name for the plot title
        filename = os.path.basename(file_path)
        
        # Create plot
        plt.figure(figsize=(10, 8))
        plt.plot(data['Longitude'], data['Latitude'], 'k.')
        plt.grid(True, linestyle='--')
        plt.xlabel('Longitude')
        plt.ylabel('Latitude')
        plt.title(f"GPS Trajectory: {filename}")
        
        # Make the plot a bit nicer
        plt.tight_layout()
        plt.show()
        
        print(f"Plotted {len(data)} GPS points from {filename}")
        
    except Exception as e:
        print(f"Error plotting data: {e}")

def main():
    # Allow user to select a file or use a specific file name
    file_path = select_file()
    
    if file_path:
        plot_trajectory(file_path)
    else:
        print("No file selected. Exiting.")

if __name__ == "__main__":
    main()

File: plot_trajectory.m
clear; close all; clc;

%% List available GPS data files
folderName = 'data';
if ~exist(folderName, 'dir')
    error('Error: Directory "%s" not found.', folderName);
end

% Get all CSV files in the data directory
fileList = dir(fullfile(folderName, 'gps*.csv'));
if isempty(fileList)
    error('No GPS data files found in the "%s" directory.', folderName);
end

% Sort by date modified (newest first)
[~, sortIdx] = sort([fileList.datenum], 'descend');
fileList = fileList(sortIdx);

% Display available files
fprintf('Available GPS data files:\n');
for i = 1:length(fileList)
    fprintf('%d. %s\n', i, fileList(i).name);
end

% Add option for latest file
fprintf('\n%d. Latest file (%s)\n', length(fileList)+1, fileList(1).name);

%% Let user select a file
choice = input(sprintf('\nSelect file to plot (1-%d), or press Enter for latest: ', length(fileList)+1));
if isempty(choice)
    choice = length(fileList)+1;  % Default to latest file
end

if choice == length(fileList)+1
    selectedFile = fileList(1);  % Latest file
elseif choice >= 1 && choice <= length(fileList)
    selectedFile = fileList(choice);
else
    error('Invalid selection.');
end

fprintf('\nPlotting trajectory from: %s\n', selectedFile.name);

%% Load and plot the selected file
filePath = fullfile(folderName, selectedFile.name);
coordinates = readmatrix(filePath);

%% plot trajectory
figure(2); clf; set(gcf, 'position', [200 100 980 420]);
subplot(1,2,1); % plot coordinates on x-y cartesian axes
latitude = coordinates(:,2);
longitude = coordinates(:,3);

plot(longitude, latitude, 'k.');
grid on; set(gca, 'gridlinestyle', '--');
xlabel('Boylam'); ylabel('Enlem');
s = 0.0001;
xmax = (1+s)*max(longitude);
xmin = (1-s)*min(longitude);
ymax = (1+s)*max(latitude);
ymin = (1-s)*min(latitude);
axis([xmin, xmax, ymin, ymax]);

subplot(1,2,2); % plot coordinates on satellite image
xlim([xmin xmax]);
ylim([ymin ymax]);

% Check if plot_google_map is available
try
    plot_google_map('MapType','satellite');
    plot(longitude, latitude, 'Marker','x', 'Color', 'Yellow');
catch
    % If plot_google_map fails, just plot normal map
    plot(longitude, latitude, 'r.');
    grid on;
    warning('Could not load satellite map. Using standard plot instead.');
end

xlabel('Longitude'); ylabel('Latitude');
axis([xmin, xmax, ymin, ymax]);

fprintf('Plotted %d GPS points from %s\n', length(latitude), selectedFile.name);